src/serializers/type_serializers/union.rs

```diff
@@ -198,24 +202,180 @@ impl TypeSerializer for UnionSerializer {
     }
 }
 
-pub struct TaggedUnionBuilder;
+#[derive(Debug, Clone)]
+pub struct TaggedUnionSerializer {
+    discriminator: Discriminator,
+    lookup: HashMap<String, CombinedSerializer>,
+    choices: Vec<CombinedSerializer>,
+    name: String,
+}
 
-impl BuildSerializer for TaggedUnionBuilder {
+impl BuildSerializer for TaggedUnionSerializer {
     const EXPECTED_TYPE: &'static str = "tagged-union";
 
     fn build(
         schema: &Bound<'_, PyDict>,
         config: Option<&Bound<'_, PyDict>>,
         definitions: &mut DefinitionsBuilder<CombinedSerializer>,
     ) -> PyResult<CombinedSerializer> {
-        let schema_choices: Bound<'_, PyDict> = schema.get_as_req(intern!(schema.py(), "choices"))?;
-        let mut choices: Vec<CombinedSerializer> = Vec::with_capacity(schema_choices.len());
+        let py = schema.py();
+        let discriminator = Discriminator::new(py, &schema.get_as_req(intern!(py, "discriminator"))?)?;
+
+        let choice_list: Bound<PyDict> = schema.get_as_req(intern!(py, "choices"))?;
+        let mut lookup: HashMap<String, CombinedSerializer> = HashMap::with_capacity(choice_list.len());
+        let mut choices: Vec<CombinedSerializer> = Vec::with_capacity(choice_list.len());
+
+        for (choice_key, choice_schema) in choice_list {
+            let serializer = CombinedSerializer::build(choice_schema.downcast()?, config, definitions).unwrap();
+            choices.push(serializer.clone());
```

Comment:
The use of `.clone()` here is not ideal, it's basically doubling the work on building the `CombinedSerializer`. Should instead wrap as `Arc<CombinedSerializer>` in both `lookup` and `choices` which will then re-use the same serializer between the two.

Inspired by this I opened #1402

Reply:
Sounds great - I have a few follow up questions here - will check in tomorrow.